File: pyapps/d3-check/d3utils/path_scanner.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Path scanner for ROSBOT and Battle.net executables.
Scans drives (D first, then others, C last) up to a fixed depth.
"""

import os
import re
import glob
from typing import List, Tuple, Optional, Callable, Dict

from providor.constants.common import BATTLE_NET_EXE_NAME, PATH_SCAN_MAX_DEPTH
from providor.constants.d3 import (
    DIABLO_III_EXE_NAME,
    ROSBOT_EXE_PATTERNS,
    ROSBOT_GAMETOOLS_BASE,
    ROSBOT_DIR_NAMESPACE_ASIA,
    ROSBOT_DIR_NAMESPACE_CN,
)
from pycore.pyfoundations.color_print import ColorPrint
from providor.providor_index import CONFIG
from d3utils.drive_order import get_fixed_drive_roots_for_scan

# Log prefix for UI (ROSBOT log receives these via ColorPrint callback)
_PATHSCAN_TAG = "[PathScan]"
# Version in path: e.g. 36.0129 -> (36, 129); same semantics as rosbot_update_manager
_VERSION_RE = re.compile(r"(\d{1,4})\.?(\d{2,5})")

# Subdirectories to skip: names starting with "." plus common cache/temp/build dirs (case-insensitive)
_SCAN_SKIP_DIRS = frozenset({
    "node_modules", "__pycache__", ".git", ".svn", ".hg", ".venv", "venv", "env",
    ".npm", ".cache", ".yarn", "tmp", "temp", "build", "dist", ".idea", ".vscode",
    "appdata", "cache", "caches", ".nuget", "packages", ".tox", ".mypy_cache",
})


def _should_skip_dir(name: str) -> bool:
    """Return True if this directory name should be skipped (dot-prefix or in skip set)."""
    if not name or name.startswith("."):
        return True
    return name.lower() in _SCAN_SKIP_DIRS


def _file_mtime(path: str) -> float:
    """Return mtime of file, or 0 on error (so path is still kept)."""
    try:
        return os.path.getmtime(path)
    except (OSError, PermissionError):
        return 0.0


def _keep_newer(
    current: Optional[Tuple[str, float]], new_path: str, new_mtime: float
) -> Optional[Tuple[str, float]]:
    """Keep the path with the later (newer) mtime. Returns (path, mtime)."""
    if current is None:
        return (new_path, new_mtime)
    return (new_path, new_mtime) if new_mtime > current[1] else current


def _scan_dir(
    root: str,
    depth: int,
    progress_callback: Optional[Callable[[str], None]] = None,
) -> Tuple[Optional[Tuple[str, float]], List[Tuple[str, float]], Optional[Tuple[str, float]]]:
    """
    Scan one directory (and subdirs up to depth).
    Returns (battlenet_path_mtime, list of (rosbot_dir, mtime), d3_path_mtime).
    When multiple matches for the same exe type are found, the one with the newest mtime is kept.
    progress_callback(current_dir) is invoked from the scan thread; do not perform UI ops in it.
    """
    battlenet: Optional[Tuple[str, float]] = None
    rosbot_list: List[Tuple[str, float]] = []
    d3: Optional[Tuple[str, float]] = None
    if depth > PATH_SCAN_MAX_DEPTH:
        return battlenet, rosbot_list, d3
    if progress_callback:
        progress_callback(root)
    try:
        entries = os.listdir(root)
    except (OSError, PermissionError):
        return battlenet, rosbot_list, d3
    for name in entries:
        full = os.path.join(root, name)
        if os.path.isfile(full):
            mtime = _file_mtime(full)
            if name == BATTLE_NET_EXE_NAME:
                battlenet = _keep_newer(battlenet, full, mtime)
            elif name == DIABLO_III_EXE_NAME:
                d3 = _keep_newer(d3, full, mtime)
            else:
                for pattern in ROSBOT_EXE_PATTERNS:
                    if glob.fnmatch.fnmatch(name, pattern):
                        rosbot_list.append((root, mtime))
                        break
        elif os.path.isdir(full):
            if _should_skip_dir(name):
                continue
            sub_bn, sub_ros, sub_d3 = _scan_dir(full, depth + 1, progress_callback)
            if sub_bn:
                battlenet = _keep_newer(battlenet, sub_bn[0], sub_bn[1])
            if sub_d3:
                d3 = _keep_newer(d3, sub_d3[0], sub_d3[1])
            rosbot_list.extend(sub_ros)
    return battlenet, rosbot_list, d3


def _get_configured_battlenet_path() -> Optional[str]:
    """Return configured Battle.net path if non-empty and file exists, else None."""
    path = (CONFIG.get("battlenet") or {}).get("battlenet_path", "").strip()
    if not path or not os.path.isfile(path):
        return None
    if os.path.basename(path) != BATTLE_NET_EXE_NAME:
        return None
    return path


def _get_configured_d3_path() -> Optional[str]:
    """Return configured D3 path if non-empty and file exists, else None."""
    path = (CONFIG.get("d3") or {}).get("d3_path", "").strip()
    if not path or not os.path.isfile(path):
        return None
    if os.path.basename(path) != DIABLO_III_EXE_NAME:
        return None
    return path


def _get_configured_ros_directory() -> Optional[str]:
    """Return configured ROSBOT directory if non-empty and dir exists, else None."""
    path = (CONFIG.get("ros_settings") or {}).get("ros_directory", "").strip()
    if not path or not os.path.isdir(path):
        return None
    return path


def are_paths_valid_for_skip_scan() -> bool:
    """
    True when Battle.net, D3, and ROSBOT are all configured and exist.
    When True, startup can skip the full path scan (user already has valid paths from a previous run).
    """
    return (
        _get_configured_battlenet_path() is not None
        and _get_configured_d3_path() is not None
        and _get_configured_ros_directory() is not None
    )


def _is_rosbot_update_convention_path(dir_path: str) -> bool:
    """True if path is under GameTools and matches {Asia|CN}_{version}\\RosBot (update convention)."""
    norm = os.path.normpath(os.path.abspath(dir_path))
    base_norm = os.path.normpath(os.path.abspath(ROSBOT_GAMETOOLS_BASE))
    if not norm.startswith(base_norm) or norm == base_norm:
        return False
    rest = norm[len(base_norm) :].lstrip(os.sep)
    parts = rest.split(os.sep)
    if len(parts) < 2:
        return False
    parent_name = parts[-2] if len(parts) >= 2 else ""
    last_part = parts[-1]
    if last_part != "RosBot":
        return False
    return (
        parent_name.startswith(ROSBOT_DIR_NAMESPACE_ASIA + "_")
        or parent_name.startswith(ROSBOT_DIR_NAMESPACE_ASIA + " ")
        or parent_name.startswith(ROSBOT_DIR_NAMESPACE_CN + "_")
        or parent_name.startswith(ROSBOT_DIR_NAMESPACE_CN + " ")
    )


def _parse_version_from_ros_path(ros_dir: str) -> Optional[Tuple[int, int]]:
    """Parse two-segment version from ROS path (e.g. .../Asia_36.0129/RosBot -> (36, 129)). Returns None if not found."""
    matches = _VERSION_RE.findall(ros_dir)
    if not matches:
        return None
    a, b = matches[-1]
    return (int(a), int(b))


def pick_best_rosbot_dir_by_region(dirs: List[str], region: Optional[str]) -> str:
    """
    Pick ROSBOT dir that matches current Battle.net region (Asia/CN). If none matches or region unknown,
    return the one with newest version (by parsed version from path).
    """
    if not dirs:
        return ""
    if region == "asia" or region == "cn":
        norm_paths = [(os.path.normpath(d), d) for d in dirs]
        for norm, orig in norm_paths:
            if region == "asia" and (ROSBOT_DIR_NAMESPACE_ASIA in norm or "亚服" in norm or "Asia" in norm):
                return orig
            if region == "cn" and (ROSBOT_DIR_NAMESPACE_CN in norm or "国服" in norm) and ROSBOT_DIR_NAMESPACE_ASIA not in norm:
                return orig
    # No match or region not asia/cn: pick by newest version
    with_ver = [(d, _parse_version_from_ros_path(d)) for d in dirs]
    with_ver = [(d, v) for d, v in with_ver if v is not None]
    if not with_ver:
        return dirs[0]
    with_ver.sort(key=lambda x: x[1], reverse=True)
    return with_ver[0][0]


def scan_for_paths(
    progress_callback: Optional[Callable[[str], None]] = None,
    include_rosbot: bool = True,
    force_scan_rosbot: bool = False,
) -> Tuple[Optional[str], List[str], Optional[str]]:
    """
    Scan from D (then other drives, C last), max PATH_SCAN_MAX_DEPTH levels.
    If configured Battle.net path exists and is valid, skip scanning for Battle.net; same for D3.
    When include_rosbot is True, ROSBOT is scanned; when False, only Battle.net and D3 are scanned.
    When force_scan_rosbot is True, always scan drives for ROSBOT (do not skip when configured), so both Asia_* and CN_* can be found for region/version matching.
    Returns (battlenet_exe_path or None, list of ROSBOT directory paths, d3_exe_path or None).
    progress_callback(current_dir) is called from the scan thread for each directory; do not perform UI operations in it.
    """
    battlenet_path: Optional[str] = None
    battlenet_mtime: float = 0.0
    d3_path: Optional[str] = None
    d3_mtime: float = 0.0
    rosbot_by_dir: Dict[str, Tuple[str, float]] = {}
    drives = get_fixed_drive_roots_for_scan()

    configured_bn = _get_configured_battlenet_path()
    if configured_bn:
        battlenet_path = configured_bn
        battlenet_mtime = _file_mtime(configured_bn)

    configured_d3 = _get_configured_d3_path()
    if configured_d3:
        d3_path = configured_d3
        d3_mtime = _file_mtime(configured_d3)

    configured_ros = _get_configured_ros_directory() if include_rosbot else None
    if configured_ros:
        rosbot_by_dir[os.path.normpath(configured_ros)] = (configured_ros, _file_mtime(configured_ros))

    ColorPrint.blue(f"{_PATHSCAN_TAG} === Scan criteria ===")
    ColorPrint.blue(f"{_PATHSCAN_TAG} Drives: {', '.join(drives) or '(none)'}, Max depth: {PATH_SCAN_MAX_DEPTH}")
    ColorPrint.blue(
        f"{_PATHSCAN_TAG} Battle.net: {'configured, skip: ' + battlenet_path if battlenet_path else 'will scan'}"
    )
    ColorPrint.blue(
        f"{_PATHSCAN_TAG} D3: {'configured, skip: ' + d3_path if d3_path else 'will scan'}"
    )
    rosbot_status = (
        "configured, will scan (force)" if (configured_ros and force_scan_rosbot and include_rosbot)
        else ("configured, skip: " + configured_ros if configured_ros and not force_scan_rosbot else ("will scan" if include_rosbot else "skip (BN+D3 only)"))
    )
    ColorPrint.blue(f"{_PATHSCAN_TAG} ROSBOT: {rosbot_status}")
    ColorPrint.blue(f"{_PATHSCAN_TAG} === Start scan ===")

    need_bn = battlenet_path is None
    need_d3 = d3_path is None
    need_rosbot = include_rosbot and (len(rosbot_by_dir) == 0 or force_scan_rosbot)

    for drive_root in drives:
        if not need_bn and not need_d3 and not need_rosbot:
            break
        if progress_callback:
            progress_callback(drive_root)
        ColorPrint.blue(f"{_PATHSCAN_TAG} Scanning {drive_root} ...")
        bn, ros, d3 = _scan_dir(drive_root, 1, progress_callback)
        if bn and (battlenet_path is None or bn[1] > battlenet_mtime):
            battlenet_path = bn[0]
            battlenet_mtime = bn[1]
            ColorPrint.green(f"{_PATHSCAN_TAG} Found Battle.net.exe: {bn[0]}")
        if d3 and (d3_path is None or d3[1] > d3_mtime):
            d3_path = d3[0]
            d3_mtime = d3[1]
            ColorPrint.green(f"{_PATHSCAN_TAG} Found Diablo III.exe: {d3[0]}")
        if include_rosbot:
            for dir_path, mtime in ros:
                norm = os.path.normpath(dir_path)
                if norm not in rosbot_by_dir or mtime > rosbot_by_dir[norm][1]:
                    rosbot_by_dir[norm] = (dir_path, mtime)
                    ColorPrint.green(f"{_PATHSCAN_TAG} Found ROSBOT: {dir_path}")
        need_bn = battlenet_path is None
        need_d3 = d3_path is None
        need_rosbot = include_rosbot and len(rosbot_by_dir) == 0
        if not need_bn and not need_d3 and not need_rosbot:
            ColorPrint.blue(f"{_PATHSCAN_TAG} Scan complete (all found).")
            break

    # Prefer update-convention paths (GameTools\{Asia|CN}_{version}\RosBot), then by mtime
    def _rosbot_sort_key(item):
        path, mtime = item
        return (0 if _is_rosbot_update_convention_path(path) else 1, -mtime)

    rosbot_sorted = sorted(rosbot_by_dir.values(), key=_rosbot_sort_key)
    rosbot_dirs = [t[0] for t in rosbot_sorted]
    bn_val = battlenet_path if battlenet_path else "(none)"
    d3_val = d3_path if d3_path else "(none)"
    ros_val = ", ".join(rosbot_dirs) if rosbot_dirs else "(none)"
    ColorPrint.blue(
        f"{_PATHSCAN_TAG} === Scan done === Battle.net: {1 if battlenet_path else 0}, D3: {1 if d3_path else 0}, ROSBOT dirs: {len(rosbot_dirs)}"
    )
    ColorPrint.blue(f"{_PATHSCAN_TAG} Battle.net: {bn_val}")
    ColorPrint.blue(f"{_PATHSCAN_TAG} D3: {d3_val}")
    ColorPrint.blue(f"{_PATHSCAN_TAG} ROSBOT dirs: {ros_val}")
    return battlenet_path, rosbot_dirs, d3_path

File: dotapps/d3check/D3CheckCore/PathScanner.cs
using System.IO;
using System.Text.RegularExpressions;
using DotCore.Foundations;

namespace DotApps.d3check.Core;

internal static class PathScanLogTag
{
    public const string Tag = "[PathScan]";
}

/// <summary>
/// Scans fixed drives for Battle.net.exe, Diablo III.exe, and ROSBOT directories.
/// Logic aligned with Python d3utils.path_scanner. Does not touch config; caller updates config and UI.
/// </summary>
public static class PathScanner
{
    private static readonly HashSet<string> SkipDirNames = new(StringComparer.OrdinalIgnoreCase)
    {
        "node_modules", "__pycache__", ".git", ".svn", ".hg", ".venv", "venv", "env",
        ".npm", ".cache", ".yarn", "tmp", "temp", "build", "dist", ".idea", ".vscode",
        "appdata", "cache", "caches", ".nuget", "packages", ".tox", ".mypy_cache",
    };

    /// <summary>
    /// Scan drives (D first, C last), max depth 6. If getConfiguredBattlenet/getConfiguredD3/getConfiguredRos return non-empty and exist, skip scanning that item unless forceScanRosbot (for ROS only).
    /// Returns (battlenetPath, rosbotDirs, d3Path). progressCallback(currentDir) is invoked from the scan; do not do UI in it.
    /// </summary>
    public static PathScanResult ScanForPaths(
        Action<string>? progressCallback = null,
        bool includeRosbot = true,
        bool forceScanRosbot = false,
        string? configuredBattlenet = null,
        string? configuredD3 = null,
        string? configuredRosDir = null)
    {
        var result = new PathScanResult();
        string? bnPath = null;
        double bnMtime = 0;
        string? d3Path = null;
        double d3Mtime = 0;
        var rosByDir = new Dictionary<string, (string path, double mtime)>(StringComparer.OrdinalIgnoreCase);

        if (!string.IsNullOrWhiteSpace(configuredBattlenet) && File.Exists(configuredBattlenet) &&
            Path.GetFileName(configuredBattlenet) == D3PathConstants.BattleNetExeName)
        {
            bnPath = configuredBattlenet;
            bnMtime = GetMtime(configuredBattlenet);
        }
        if (!string.IsNullOrWhiteSpace(configuredD3) && File.Exists(configuredD3) &&
            Path.GetFileName(configuredD3) == D3PathConstants.DiabloIIIExeName)
        {
            d3Path = configuredD3;
            d3Mtime = GetMtime(configuredD3);
        }
        if (!string.IsNullOrWhiteSpace(configuredRosDir) && Directory.Exists(configuredRosDir))
            rosByDir[NormalizePath(configuredRosDir)] = (configuredRosDir, GetMtime(configuredRosDir));

        string[] drives = DriveOrder.GetFixedDriveRootsForScan();
        ColorPrinter.Blue($"{PathScanLogTag.Tag} === Scan criteria ===");
        ColorPrinter.Blue($"{PathScanLogTag.Tag} Drives: {(drives.Length > 0 ? string.Join(", ", drives) : "(none)")}, Max depth: {D3PathConstants.PathScanMaxDepth}");
        ColorPrinter.Blue($"{PathScanLogTag.Tag} Battle.net: {(bnPath != null ? "configured, skip: " + bnPath : "will scan")}");
        ColorPrinter.Blue($"{PathScanLogTag.Tag} D3: {(d3Path != null ? "configured, skip: " + d3Path : "will scan")}");
        string rosStatus = includeRosbot
            ? (rosByDir.Count > 0 && forceScanRosbot ? "configured, will scan (force)" : (rosByDir.Count > 0 ? "configured, skip: " + rosByDir.Values.First().path : "will scan"))
            : "skip (BN+D3 only)";
        ColorPrinter.Blue($"{PathScanLogTag.Tag} ROSBOT: {rosStatus}");
        ColorPrinter.Blue($"{PathScanLogTag.Tag} === Start scan ===");

        bool needBn = bnPath == null;
        bool needD3 = d3Path == null;
        bool needRos = includeRosbot && (rosByDir.Count == 0 || forceScanRosbot);

        foreach (string driveRoot in drives)
        {
            if (!needBn && !needD3 && !needRos) break;
            progressCallback?.Invoke(driveRoot);
            ColorPrinter.Blue($"{PathScanLogTag.Tag} Scanning {driveRoot} ...");

            ScanDir(driveRoot, 1, progressCallback, ref bnPath, ref bnMtime, ref d3Path, ref d3Mtime, rosByDir, includeRosbot);

            needBn = bnPath == null;
            needD3 = d3Path == null;
            needRos = includeRosbot && rosByDir.Count == 0;
            if (!needBn && !needD3 && !needRos)
            {
                ColorPrinter.Blue($"{PathScanLogTag.Tag} Scan complete (all found).");
                break;
            }
        }

        result.BattlenetPath = bnPath;
        result.D3Path = d3Path;
        // 1:1 Python: prefer update-convention paths (GameTools\{Asia|CN}_{version}\RosBot), then by mtime descending
        result.RosbotDirs = rosByDir.Values
            .OrderBy(x => IsRosbotUpdateConventionPath(x.path) ? 0 : 1)
            .ThenByDescending(x => x.mtime)
            .Select(x => x.path)
            .ToList();

        ColorPrinter.Blue($"{PathScanLogTag.Tag} === Scan done === Battle.net: {(bnPath != null ? 1 : 0)}, D3: {(d3Path != null ? 1 : 0)}, ROSBOT dirs: {result.RosbotDirs.Count}");
        ColorPrinter.Blue($"{PathScanLogTag.Tag} Battle.net: {bnPath ?? "(none)"}");
        ColorPrinter.Blue($"{PathScanLogTag.Tag} D3: {d3Path ?? "(none)"}");
        ColorPrinter.Blue($"{PathScanLogTag.Tag} ROSBOT dirs: {(result.RosbotDirs.Count > 0 ? string.Join(", ", result.RosbotDirs) : "(none)")}");

        return result;
    }

    private static void ScanDir(
        string root,
        int depth,
        Action<string>? progressCallback,
        ref string? battlenetPath,
        ref double battlenetMtime,
        ref string? d3Path,
        ref double d3Mtime,
        Dictionary<string, (string path, double mtime)> rosByDir,
        bool includeRosbot)
    {
        if (depth > D3PathConstants.PathScanMaxDepth) return;
        progressCallback?.Invoke(root);

        if (!Directory.Exists(root)) return;
        string[] entries;
        try
        {
            entries = Directory.GetFileSystemEntries(root);
        }
        catch
        {
            return;
        }

        foreach (string full in entries)
        {
            if (File.Exists(full))
            {
                string name = Path.GetFileName(full);
                double mtime = GetMtime(full);
                if (name == D3PathConstants.BattleNetExeName)
                    KeepNewer(ref battlenetPath, ref battlenetMtime, full, mtime);
                else if (name == D3PathConstants.DiabloIIIExeName)
                    KeepNewer(ref d3Path, ref d3Mtime, full, mtime);
                else if (includeRosbot && MatchesRosbot(name))
                {
                    string dir = Path.GetDirectoryName(full) ?? full;
                    string norm = NormalizePath(dir);
                    if (!rosByDir.TryGetValue(norm, out var existing) || mtime > existing.mtime)
                        rosByDir[norm] = (dir, mtime);
                }
            }
            else if (Directory.Exists(full))
            {
                string name = Path.GetFileName(full);
                if (string.IsNullOrEmpty(name) || name[0] == '.' || SkipDirNames.Contains(name))
                    continue;
                ScanDir(full, depth + 1, progressCallback, ref battlenetPath, ref battlenetMtime, ref d3Path, ref d3Mtime, rosByDir, includeRosbot);
            }
        }
    }

    private static bool MatchesRosbot(string fileName)
    {
        foreach (string pattern in D3PathConstants.RosbotExePatterns)
        {
            if (MatchesGlob(fileName, pattern)) return true;
        }
        return false;
    }

    private static bool MatchesGlob(string name, string pattern)
    {
        string regex = "^" + Regex.Escape(pattern).Replace("\\*", ".*").Replace("\\?", ".") + "$";
        return Regex.IsMatch(name, regex, RegexOptions.IgnoreCase);
    }

    private static void KeepNewer(ref string? current, ref double currentMtime, string newPath, double newMtime)
    {
        if (current == null || newMtime > currentMtime)
        {
            current = newPath;
            currentMtime = newMtime;
        }
    }

    private static double GetMtime(string path)
    {
        try
        {
            return File.GetLastWriteTimeUtc(path).Ticks;
        }
        catch
        {
            return 0;
        }
    }

    private static string NormalizePath(string path) => Path.GetFullPath(path).TrimEnd(Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar);

    /// <summary>True if path is under GameTools and matches {Asia|CN}_{version}\RosBot (update convention). 1:1 Python _is_rosbot_update_convention_path.</summary>
    private static bool IsRosbotUpdateConventionPath(string dirPath)
    {
        try
        {
            string norm = Path.GetFullPath(dirPath).TrimEnd(Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar);
            string baseNorm = Path.GetFullPath(D3PathConstants.RosbotGameToolsBase).TrimEnd(Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar);
            if (string.IsNullOrEmpty(baseNorm) || !norm.StartsWith(baseNorm, StringComparison.OrdinalIgnoreCase) || norm.Length <= baseNorm.Length)
                return false;
            string rest = norm.Substring(baseNorm.Length).TrimStart(Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar);
            string[] parts = rest.Split(Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar);
            if (parts.Length < 2) return false;
            string parentName = parts[^2];
            string lastPart = parts[^1];
            if (!string.Equals(lastPart, D3PathConstants.RosbotFinalDirName, StringComparison.OrdinalIgnoreCase))
                return false;
            return parentName.StartsWith(D3PathConstants.RosbotDirNamespaceAsia + "_", StringComparison.OrdinalIgnoreCase)
                || parentName.StartsWith(D3PathConstants.RosbotDirNamespaceAsia + " ", StringComparison.OrdinalIgnoreCase)
                || parentName.StartsWith(D3PathConstants.RosbotDirNamespaceCn + "_", StringComparison.OrdinalIgnoreCase)
                || parentName.StartsWith(D3PathConstants.RosbotDirNamespaceCn + " ", StringComparison.OrdinalIgnoreCase);
        }
        catch
        {
            return false;
        }
    }
}
